course6/toyapp-adam/math/matrix.cc

```diff
@@ -0,0 +1,135 @@
+#include "matrix.h"
+
+#include <cmath>
+
+namespace Math {
+
+namespace {
+	constexpr auto PI = std::atan(1.0) * 4.0;
+}
+
+Matrix44 Matrix44::Zero()
+{
+	Matrix44 m;
+
+	for(unsigned i = 0; i < 16; ++i)
+		m.mMatrix[i] = 0.0f;
+
+	return m;
+}
+
+Matrix44 Matrix44::Identity()
+{
+	Matrix44 m = Zero();
+
+	m.mMatrix[ 0] = 1.0f;
+	m.mMatrix[ 5] = 1.0f;
+	m.mMatrix[10] = 1.0f;
+	m.mMatrix[15] = 1.0f;
+
+	return m;
+}
+
+Matrix44 Matrix44::Translation(Vector3D position)
+{
+	Matrix44 m = Identity();
+
+	m.mMatrix[TRANSLATE_X] = position.x;
+	m.mMatrix[TRANSLATE_Y] = position.y;
+	m.mMatrix[TRANSLATE_Z] = position.z;
+
+	return m;
+}
+
+Matrix44 Matrix44::Rotation(Vector3D yawPitchRoll)
+{
+	Matrix44 m = Zero();
+
+	const float ch = std::cos(yawPitchRoll.x);
+	const float sh = std::sin(yawPitchRoll.x);
+	const float cp = std::cos(yawPitchRoll.y);
+	const float sp = std::sin(yawPitchRoll.y);
+	const float cb = std::cos(yawPitchRoll.z);
+	const float sb = std::sin(yawPitchRoll.z);
+
+	const float shsp = sh * sp;
+	const float chsp = ch * sp;
+
+	m.mMatrix[ 0] =  ch * cb + shsp * sb;
+	m.mMatrix[ 1] = -ch * sb + shsp * cb;
+	m.mMatrix[ 2] =  sh * cp;
+	m.mMatrix[ 4] =  sb * cp;
+	m.mMatrix[ 5] =  cb * cp;
+	m.mMatrix[ 6] = -sp;
+	m.mMatrix[ 8] = -sh * cb + chsp * sb;
+	m.mMatrix[ 9] =  sb * sh + chsp * cb;
+	m.mMatrix[10] =  ch * cp;
+	m.mMatrix[15] =  1.0f;
+
+	return m;
+}
+
+Matrix44 Matrix44::Scale(float size)
+{
+	Matrix44 m = Identity();
+
+	m.mMatrix[SCALE_X] = size;
+	m.mMatrix[SCALE_Y] = size;
+	m.mMatrix[SCALE_Z] = size;
+
+	return m;
+}
+
+Matrix44 Matrix44::Perspective(float fieldOfViewDegrees, float aspectRatio)
+{
+	Matrix44 m = Zero();
+
+	m.mMatrix[SCALE_Y] = 1.0f / std::tan((PI / 180.0f) * fieldOfViewDegrees * 0.5f);
+	m.mMatrix[SCALE_X] = m.mMatrix[SCALE_Y] / aspectRatio;
+	m.mMatrix[11] = +1.0f;
+	m.mMatrix[14] = -1.0f;
+
+	return m;
+}
+
+Matrix44 Matrix44::operator*(const Matrix44& rhs) const
+{
+	constexpr unsigned ROWS = 4;
+	constexpr unsigned COLS = 4;
```

Comment:
I think this is a stylistic overkill; you could just use literal 4 in the code.
The constexprs have no real meaning in this case either.

Reply:
I was under the impression it's preferable to use constexpr if possible. It's not something I'm very sure about though, so I can stop doing this. :)